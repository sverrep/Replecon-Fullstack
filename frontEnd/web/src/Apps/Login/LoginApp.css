.logo-h1{
  text-align: center;
  color:#0d6efd;
  padding-bottom: 2%;
}

.login-container{
  display: flex;
  justify-content: center;
  width: 50%;
}

.vertical-center {
  margin: 0;
  position: absolute;
  top: 50%;
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.login-h3{
  text-align: center;
}

.login-btns{
  margin-right: 5px;
  margin-left: 5px;
  margin-bottom: 15px;
  width: 60%;
}

.login-btns-row{
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
}
.footerlog {
  position: fixed;
  text-align: center;
  padding: 5px 5px 0px 5px;
  bottom: 0;
  width: 100%;
  /* Height of the footer*/ 
  height: 30px;
  background: rgb(235, 235, 235);

.error-message{
  text-align: center;
  color:red;
}