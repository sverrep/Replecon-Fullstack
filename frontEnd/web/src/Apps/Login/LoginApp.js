import axios from 'axios';
import React from 'react';
import { Redirect } from "react-router-dom";
import getIP from '../../settings.js';
import './LoginApp.css';
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Form from 'react-bootstrap/Form'
import Button from 'react-bootstrap/Button'

export default class LoginApp extends React.Component {
  constructor(props) {
    super(props);
    this.state = { email: '', password: '', redirect_student_profile: false, redirect_student_signup: false, redirect_teacher_signup: false, error: '' };
    this.handleChange = this.handleChange.bind(this);
    this.handleLogin = this.handleLogin.bind(this);
    this.handleStudentSignUpRedirect = this.handleStudentSignUpRedirect.bind(this);
    this.handleTeacherSignUpRedirect = this.handleTeacherSignUpRedirect.bind(this);
  }

  render() {
    if (this.state.redirect_student_signup){
      return(
        <Redirect to={{pathname: '/SignUp', state: { role: "Student" }}}></Redirect>
      )
    }
    else if (this.state.redirect_teacher_signup){
      return(
        <Redirect to={{pathname: '/SignUp', state: { role: "Teacher" }}}></Redirect>
      )
    }
    else if(this.state.redirect_student_profile){
      return(
        <Redirect to={{pathname: '/Profile', state: { email: this.state.email, role: "Student" }}}></Redirect>
      )
    }
    else if(this.state.redirect_teacher_profile){
      return(
        <Redirect to={{pathname: '/Profile', state: { email: this.state.email,  role: "Teacher" }}}></Redirect>
      )
    }
    else{
      return (
        <div>
        <Container className="login-container">
          <div className="vertical-center">
            <Row>
              <h1 className="logo-h1">REPLECON</h1>
              <h3 className="login-h3">Login</h3>
            </Row>
            <Row>
              <Form.Group className="mb-3" controlId="email">
                <Form.Label>Email address</Form.Label>
                <Form.Control type="email" placeholder="Enter email" onChange={this.handleChange} />
              </Form.Group>
              <Form.Group className="mb-3" controlId="password">
                <Form.Label>Password</Form.Label>
                <Form.Control type="password" placeholder="Password" onChange={this.handleChange}/>
              </Form.Group>
            </Row>
            <p className="error-message">{this.state.error}</p>
            <div className="login-btns-row">
              <Button variant="primary" className="login-btns" onClick={() => this.handleLogin()}> Login </Button>
            </div>
            <div className= "login-btns-row">
              <Button variant="secondary" className="login-btns" onClick={this.handleStudentSignUpRedirect} > Student Sign Up </Button>
              <Button variant="secondary" className="login-btns" onClick={this.handleTeacherSignUpRedirect} >Teacher Sign Up</Button>
            </div>
            
          </div>
        </Container>
        <div className='footerlog'>
          <p>SPSO Solutions Â©</p>
        </div>
        </div>
      );
    }
  }

  handleStudentSignUpRedirect(){
    this.setState({ redirect_student_signup : true })
  }

  handleTeacherSignUpRedirect(){
    this.setState({ redirect_teacher_signup : true })
  }

  handleChange(e) {
    const field = e.target.id
    this.setState({ [field] : e.target.value });
  }

  validateData() 
  {
    var reg = /^\w+([-]?\w+)*@\w+([-]?\w+)*(\.\w\w+)+$/;
    if (reg.test(this.state.email))
    {
      if(this.state.password.length === 0){
          this.setState({error: "Password needs to be longer", showError: true})
      }
      else{
        return true
      }
    }
    else
    {
      this.setState({error: "Email is invalid", showError: true})
    } 
  }

  handleLogin() {
    if (this.validateData()) {
      const payload = { username: this.state.email, password: this.state.password } 
      axios.post(getIP()+'/auth/login/', payload)
      .then(response => {
        const { token } = response.data;
        axios.defaults.headers.common.Authorization = `Token ${token}`;
        axios.get(getIP()+'/teachers/isTeacher/')
        .then(response => {
          if(response.data === true) {
            this.setState({ redirect_teacher_profile : true})
            return
          }
          else {
            this.setState({ redirect_student_profile : true})
            return
          }
        })
        .catch(error => console.log(error));
      })
      .catch(error => console.log(error));
      this.setState({error: "User Not Found"})
      return
    }
    else {
      return
    }
    
  }
}
