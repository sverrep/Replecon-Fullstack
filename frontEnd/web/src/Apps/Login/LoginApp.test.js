import React from 'react';
import ReactDOM from 'react-dom';
import LoginApp from './LoginApp.js'
import axios from 'axios'
jest.mock('axios', () => {
    return {
      post: jest.fn(() => Promise.resolve({ data: {} })),
    };
  });

it('should fail validateData due to email', () => {
    const wrapper = shallow(<LoginApp />)
    wrapper.instance().state.email = "student"
    wrapper.instance().state.password = "12345"
    const condition = wrapper.instance().validateData()
    expect(wrapper.instance().state.error).toBe("Email is invalid")
    expect(condition).toBe(undefined)
});

it('should fail validateData due to password', () => {
    const wrapper = shallow(<LoginApp />)
    wrapper.instance().state.email = "student@mail.com"
    wrapper.instance().state.password = ""
    const condition = wrapper.instance().validateData()
    expect(wrapper.instance().state.error).toBe("Password needs to be longer")
    expect(condition).toBe(undefined)
});

it('should pass validateData', () => {
    const wrapper = shallow(<LoginApp />)
    wrapper.instance().state.email = "student@mail.com"
    wrapper.instance().state.password = "12345"
    const condition = wrapper.instance().validateData()
    expect(wrapper.instance().state.error).toBe("")
    expect(condition).toBe(true)
});

it('should change email', () => {
    const wrapper = shallow(<LoginApp />)
    const event = {
        preventDefault() {},
        target: {id: "email", value: "teacher@mail.com"}
    }
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.email).toBe("teacher@mail.com")

});

it('should change password', () => {
    const wrapper = shallow(<LoginApp />)
    const event = {
        preventDefault() {},
        target: {id: "password", value: "54321"}
    }
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.password).toBe("54321")

});

it('should change nothing', () => {
    const wrapper = shallow(<LoginApp />)
    const event = {
        preventDefault() {},
        target: {id: "", value: "value"}
    }
    wrapper.instance().state.email = "student@mail.com"
    wrapper.instance().state.password = "12345"
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.email).toBe("student@mail.com")
    expect(wrapper.instance().state.password).toBe("12345")

});

it('should change nothing', () => {
    const wrapper = shallow(<LoginApp />)
    const event = {
        preventDefault() {},
        target: {id: "", value: "value"}
    }
    wrapper.instance().state.email = "student@mail.com"
    wrapper.instance().state.password = "12345"
    wrapper.instance().handleChange(event)
    expect(wrapper.instance().state.email).toBe("student@mail.com")
    expect(wrapper.instance().state.password).toBe("12345")

});

it('should call set student_profile_redirect', async () => {
    const wrapper = shallow(<LoginApp />)
    wrapper.instance().validateData = jest.fn()
    wrapper.instance().validateData.mockReturnValueOnce(true)
    wrapper.instance().state.email = "student@mail.com"
    wrapper.instance().state.password = "12345"
    wrapper.instance().state.token = ""
    const payload = {
        username: wrapper.instance().state.email,
        password: wrapper.instance().state.password
    }
    const users = mockResponseData(true, ['John Doe', 'Charles']);
    axios.post.mockResolvedValue(() => {
        return Promise.resolve(users);
      });

    const result = await wrapper.instance().handleLogin()
    console.log(result)
    expect(wrapper.instance().validateData).toHaveBeenCalledTimes(1)
    expect(axios.post).toHaveBeenCalledTimes(1)
    expect(axios.post).toHaveBeenCalledWith('https://mythical-mason-324813.ey.r.appspot.com//auth/login/', payload)
    expect(result).toBe(users)


});



const mockResponseData = jest.fn((success, payload) => {
    return {
      data: {
        result: {
          success,
          payload,
        },
      },
    };
  });