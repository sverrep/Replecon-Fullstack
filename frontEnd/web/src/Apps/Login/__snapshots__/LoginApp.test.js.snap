// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginApp Testing component render testing Login page renders correctly 1`] = `
Array [
  <div>
    <Container
      className="login-container"
      fluid={false}
    >
      <div
        className="vertical-center"
      >
        <Row>
          <h1
            className="logo-h1"
          >
            REPLECON
          </h1>
          <h3
            className="login-h3"
          >
            Login
          </h3>
        </Row>
        <Row>
          <FormGroup
            className="mb-3"
            controlId="email"
          >
            <FormLabel
              column={false}
              visuallyHidden={false}
            >
              Email address
            </FormLabel>
            <FormControl
              onChange={[Function]}
              placeholder="Enter email"
              type="email"
            />
          </FormGroup>
          <FormGroup
            className="mb-3"
            controlId="password"
          >
            <FormLabel
              column={false}
              visuallyHidden={false}
            >
              Password
            </FormLabel>
            <FormControl
              onChange={[Function]}
              placeholder="Password"
              type="password"
            />
          </FormGroup>
        </Row>
        <p
          className="error-message"
        >
          
        </p>
        <div
          className="login-btns-row"
        >
          <Button
            active={false}
            className="login-btns"
            disabled={false}
            onClick={[Function]}
            variant="primary"
          >
             Login 
          </Button>
        </div>
        <div
          className="login-btns-row"
        >
          <Button
            active={false}
            className="login-btns"
            disabled={false}
            onClick={[Function]}
            variant="secondary"
          >
             Student Sign Up 
          </Button>
          <Button
            active={false}
            className="login-btns"
            disabled={false}
            onClick={[Function]}
            variant="secondary"
          >
            Teacher Sign Up
          </Button>
        </div>
      </div>
    </Container>
    <div
      className="footerlog"
    >
      <p>
        SPSO Solutions Â©
      </p>
    </div>
  </div>,
]
`;
